examples/helloworld_client.rs

```diff
@@ -0,0 +1,106 @@
+extern crate env_logger;
+extern crate futures;
+extern crate http;
+#[macro_use]
+extern crate prost_derive;
+extern crate tokio_connect;
+extern crate tokio_core;
+extern crate tower;
+extern crate tower_grpc;
+extern crate tower_h2;
+
+use futures::Future;
+use tokio_connect::Connect;
+use tokio_core::net::TcpStream;
+use tokio_core::reactor::{Core, Handle};
+use tower::NewService;
+use tower_grpc::codegen::client::*;
+
+use std::net::SocketAddr;
+
+struct Conn(SocketAddr, Handle);
+
+impl Connect for Conn {
+    type Connected = TcpStream;
+    type Error = ::std::io::Error;
+    type Future = Box<Future<Item = TcpStream, Error = ::std::io::Error>>;
+
+    fn connect(&self) -> Self::Future {
+        let c = TcpStream::connect(&self.0, &self.1)
+            .and_then(|tcp| tcp.set_nodelay(true).map(move |_| tcp));
+        Box::new(c)
+    }
+}
+
+/// The request message containing the user's name.
+#[derive(Clone, Debug, PartialEq, Message)]
+pub struct HelloRequest {
+    #[prost(string, tag="1")]
+    pub name: String,
+}
+
+/// The response message containing the greetings
+#[derive(Clone, Debug, PartialEq, Message)]
+pub struct HelloReply {
+    #[prost(string, tag="1")]
+    pub message: String,
+}
+
+#[derive(Debug)]
+pub struct Greeter<T> {
+    /// The inner HTTP/2.0 service
+    inner: grpc::Grpc<T>,
+}
+
+use tower_grpc::client::Encodable;
+
+impl<T> Greeter<T>
+where T: tower_h2::HttpService,
+{
+    pub fn new(inner: T, uri: http::Uri) -> Self {
+        let inner = grpc::Builder::new()
+            .uri(uri)
+            .build(inner);
+
+        Greeter { inner }
+    }
+
+    pub fn poll_ready(&mut self) -> futures::Poll<(), grpc::Error<T::Error>> {
+        self.inner.poll_ready()
+    }
+
+    pub fn say_hello(&mut self, request: grpc::Request<HelloRequest>)
+        -> grpc::unary::ResponseFuture<HelloReply, T::Future, T::ResponseBody>
+    where grpc::unary::Once<HelloRequest>: Encodable<T::RequestBody>,
+    {
+        let path = http::uri::PathAndQuery::from_static("/helloworld.Greeter/SayHello");
+        self.inner.unary(request, path)
```

Comment:
Why isn't the path part of the `grpc::Request`? It seems more appropriate there...

Reply:
I removed it because that information is never supplied by the user (it is set by the client and server lib) and the user can get that info by knowing which grpc fn they are calling / being called from. 

Even if the user could read the value, they don’t know what the value should be when creating the grpc request struct which means it would need to be set to None (using a default value would mean if the request were passed on to a different fn. and that Fn read the default value, they wouldn’t have anyway to know it was a default value and not a real value)

Removing it also means the struct size can be reduced.

Overall it made things much cleaner.

Reply:
I'm not sure I agree, it's the same as someone needing to know the `Uri` when they want to create an `http::Request`. As for the size of the struct, the `Uri` needs to be passed as a required separate argument, so I don't see a gain.

Reply:
I don't see how it is the same as with HTTP. In the HTTP case you *have* to set the URI to send a request. An HTTP client has no way of knowing what the URI should be.

In the case of the gRPC client (at least the protobuf one), you would never set the URI, the URI is going to be set internally to the client. In fact, the client is not expected to even know what the request URI should be as it is an internal detail of codegen. All the issues that I outlined above are not true for an HTTP client.

Either way, not exposing a URI is the more conservative option here, if I am wrong, we can always add it back later.